### PR TITLE
Fix flaky ORC test by using NullMemoryManager

### PR DESCRIPTION
**Pull request by haozhun**

### CODE REVIEW

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 1]
```diff
@@ -418,7 +432,9 @@ public static FileSplit createTestFile(String filePath,
             throws Exception
     {
         // filter out partition keys, which are not written to the file
-        testColumns = ImmutableList.copyOf(filter(testColumns, not(TestColumn::isPartitionKey)));
+        testColumns = testColumns.stream()
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: This seems unrelated?

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 2]
```diff
@@ -426,27 +442,18 @@ public static FileSplit createTestFile(String filePath,
         Properties tableProperties = new Properties();
         tableProperties.setProperty("columns", Joiner.on(',').join(transform(testColumns, TestColumn::getName)));
         tableProperties.setProperty("columns.types", Joiner.on(',').join(transform(testColumns, TestColumn::getType)));
-        tableProperties.setProperty("orc.stripe.size", "1200000");
-        serDe.initialize(new Configuration(), tableProperties);
+        // Setting orc.stripe.size property here does not help because flushStripe is never called with NullMemoryManager regardless of the configuration.
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: This comment doesn't seem relevant anymore

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 3]
```diff
@@ -466,6 +473,9 @@ public static FileSplit createTestFile(String filePath,
 
                 Writable record = serDe.serialize(row, objectInspector);
                 recordWriter.write(record);
+                if (rowNumber % STRIPE_ROWS == STRIPE_ROWS - 1) {
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: Add parens

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 4]
```diff
@@ -473,11 +483,55 @@ public static FileSplit createTestFile(String filePath,
...
 
         Path path = new Path(filePath);
-        path.getFileSystem(new Configuration()).setVerifyChecksum(true);
+        path.getFileSystem(CONFIGURATION).setVerifyChecksum(true);
         File file = new File(filePath);
         return new FileSplit(path, 0, file.length(), new String[0]);
     }
 
+    private static void flushStripe(FileSinkOperator.RecordWriter recordWriter)
+            throws ClassNotFoundException, NoSuchFieldException, IllegalAccessException, NoSuchMethodException, InvocationTargetException
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: ReflectiveOperationException

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 5]
```diff
@@ -473,11 +483,55 @@ public static FileSplit createTestFile(String filePath,
...
+        Method flushStripe = WriterImpl.class.getDeclaredMethod("flushStripe");
+        flushStripe.setAccessible(true);
+        flushStripe.invoke(writer);
+    }
+
+    private static FileSinkOperator.RecordWriter createRecordWriter(Path target, Configuration conf)
+            throws IOException
+    {
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(FileSystem.class.getClassLoader())) {
+            // Setting stripeSize here does not help because flushStripe is never called with NullMemoryManager regardless of the configuration.
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: Remove this comment

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 6]
```diff
@@ -473,11 +483,55 @@ public static FileSplit createTestFile(String filePath,
...
-        path.getFileSystem(new Configuration()).setVerifyChecksum(true);
+        path.getFileSystem(CONFIGURATION).setVerifyChecksum(true);
         File file = new File(filePath);
         return new FileSplit(path, 0, file.length(), new String[0]);
     }
 
+    private static void flushStripe(FileSinkOperator.RecordWriter recordWriter)
+            throws ClassNotFoundException, NoSuchFieldException, IllegalAccessException, NoSuchMethodException, InvocationTargetException
+    {
+        Field writerField = OrcOutputFormat.class.getClassLoader().loadClass(OrcOutputFormat.class.getName() + "$OrcRecordWriter").getDeclaredField("writer");
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: This Class is duplicated below

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 7]
```diff
@@ -473,11 +483,55 @@ public static FileSplit createTestFile(String filePath,
...
+            try {
+                return WRITER_CONSTRUCTOR.newInstance(target, options);
+            }
+            catch (Throwable e) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    private static Constructor<? extends FileSinkOperator.RecordWriter> getOrcWriterConstructor()
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: This code seems duplicated. Can we pull it into the Orc module?

**Comment by haozhun** [comment]
Addressed comments

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 8]
```diff
@@ -97,8 +109,13 @@
 
 public class TestOrcPageSourceMemoryTracking
 {
+    private static final String ORC_OUTPUT_FORMAT_ORC_RECORD_READER = OrcOutputFormat.class.getName() + "$OrcRecordWriter";
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: Call this `ORC_RECORD_WRITER`

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 9]
```diff
@@ -97,8 +109,13 @@
...
+    private static final Constructor<? extends FileSinkOperator.RecordWriter> WRITER_CONSTRUCTOR = getOrcWriterConstructor();
+    private static final Configuration CONFIGURATION = new Configuration();
     private static final TypeManager TYPE_MANAGER = new TypeRegistry();
     private static final int NUM_ROWS = 50000;
+    private static final int STRIPE_ROWS = 20000;
+    private static Class<?> aClass;
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: This seems unused (and is also a bad variable name)

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 10]
```diff
@@ -429,27 +446,17 @@ public static FileSplit createTestFile(String filePath,
...
 
-        FileSinkOperator.RecordWriter recordWriter = outputFormat.getHiveRecordWriter(
-                jobConf,
-                new Path(filePath),
-                Text.class,
-                compressionCodec != null,
-                tableProperties,
-                () -> { }
-        );
+        FileSinkOperator.RecordWriter recordWriter = createRecordWriter(new Path(filePath), CONFIGURATION);
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: Static import `RecordWriter`

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 11]
```diff
@@ -476,11 +486,59 @@ public static FileSplit createTestFile(String filePath,
...
-        path.getFileSystem(new Configuration()).setVerifyChecksum(true);
+        path.getFileSystem(CONFIGURATION).setVerifyChecksum(true);
         File file = new File(filePath);
         return new FileSplit(path, 0, file.length(), new String[0]);
     }
 
+    private static void flushStripe(FileSinkOperator.RecordWriter recordWriter)
+    {
+        try {
+            Field writerField = OrcOutputFormat.class.getClassLoader().loadClass(ORC_OUTPUT_FORMAT_ORC_RECORD_READER).getDeclaredField("writer");
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: Wrap:
```java
Field writerField = OrcOutputFormat.class.getClassLoader()
        .loadClass(ORC_RECORD_WRITER).getDeclaredField("writer");

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 12]
```diff
@@ -476,11 +486,59 @@ public static FileSplit createTestFile(String filePath,
...
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    private static Constructor<? extends FileSinkOperator.RecordWriter> getOrcWriterConstructor()
+    {
+        try {
+            Constructor<? extends FileSinkOperator.RecordWriter> constructor =
+                    OrcOutputFormat.class.getClassLoader()
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: Format like this:
```java
Constructor<? extends RecordWriter> constructor = OrcOutputFormat.class.getClassLoader()
        .loadClass(ORC_RECORD_WRITER)
        .asSubclass(RecordWriter.class)
        .getDeclaredConstructor(Path.class, OrcFile.WriterOptions.class);

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 13]
```diff
@@ -476,11 +486,59 @@ public static FileSplit createTestFile(String filePath,
...
+        catch (ReflectiveOperationException e) {
+            throw Throwables.propagate(e);
+        }
+    }
+
+    private static FileSinkOperator.RecordWriter createRecordWriter(Path target, Configuration conf)
+            throws IOException
+    {
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(FileSystem.class.getClassLoader())) {
+            OrcFile.WriterOptions options = new OrcWriterOptions(conf)
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: Static import `WriterOptions`

**Comment by electrum** [comment]
Looks good
